File: src/Services/Campus/Campus.Domain/Entities/Room.cs
using System;

namespace Campus.Domain.Entities;

public class Room
{
    public Guid Id { get; private set; }
    public Guid BuildingId { get; private set; }
    public Guid CategoryId { get; private set; }
    
    public string Name { get; private set; } = string.Empty;
    public int Floor { get; private set; }
    public string? Number { get; private set; }
    
    public Building Building { get; private set; } = null!;
    public Category Category { get; private set; } = null!;

    public Room(string name, int floor, string? number, Building building, Category category)
    {
        this.Id = Guid.NewGuid();
        this.Name = name;
        this.Floor = floor;
        this.Number = number;
        
        this.Building = building;
        this.BuildingId = building.Id;
        
        this.Category = category;
        this.CategoryId = category.Id;
    }

    public Room(string name, int floor, string? number, Guid buildingId, Guid categoryId)
    {
        this.Id = Guid.NewGuid();
        this.Name = name;
        this.Floor = floor;
        this.Number = number;
        this.BuildingId = buildingId;
        this.CategoryId = categoryId;
    }

    public void Update(string name, int floor, string? number, Guid buildingId, Guid categoryId)
    {
        this.Name = name;
        this.Floor = floor;
        this.Number = number;
        this.BuildingId = buildingId;
        this.CategoryId = categoryId;
    }

    protected Room() { }
}

File: uct_map/lib/domain/entities/room.dart
/// Entidad Sala u oficina dentro de un Edificio (ms.svg - Campus Service).
class Room {
  final String id;
  final String buildingId;
  final String categoryId;
  final String name;
  final int floor;
  final String? number;

  const Room({
    required this.id,
    required this.buildingId,
    required this.categoryId,
    required this.name,
    required this.floor,
    this.number,
  });

  factory Room.fromJson(Map<String, dynamic> json) {
    return Room(
      id: (json['id'] ?? json['Id'] ?? '').toString(),
      buildingId: (json['buildingId'] ?? json['BuildingId'] ?? '').toString(),
      categoryId: (json['categoryId'] ?? json['CategoryId'] ?? '').toString(),
      name: (json['name'] ?? json['Name'] ?? '').toString(),
      floor: (json['floor'] ?? json['Floor'] ?? 1) as int,
      number: json['number']?.toString() ?? json['Number']?.toString(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'buildingId': buildingId,
        'categoryId': categoryId,
        'name': name,
        'floor': floor,
        if (number != null) 'number': number,
      };
}
